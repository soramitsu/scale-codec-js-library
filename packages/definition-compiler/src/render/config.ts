import { InjectionKey, inject, provide } from 'vue';

const key: InjectionKey<Config> = Symbol('render config');

export interface Config {
    importLib: string;
}

export function useConfig(): Config {
    const val = inject(key);
    if (!val) throw new Error('no config');
    return val;
}

export function provideConfig(conf: Config): void {
    provide(key, conf);
}
