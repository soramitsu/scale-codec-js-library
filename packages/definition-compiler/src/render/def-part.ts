export type DefPart = 'ty-decoded' | 'ty-encodable' | 'fn-decode' | 'fn-encode';

export const DefPartSuffixMap: { [K in DefPart]: string } = {
    'fn-decode': '_decode',
    'fn-encode': '_encode',
    'ty-decoded': '_Decoded',
    'ty-encodable': '_Encodable',
};

export function partToSuffix(val: DefPart) {
    return DefPartSuffixMap[val];
}
