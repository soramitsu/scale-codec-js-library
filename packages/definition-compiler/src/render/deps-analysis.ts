import { Set, Map, List } from 'immutable'
import Graph from 'tarjan-graph'

export type AdjacencyMap = Map<string, Set<string>>

/**
 * This function is an implementation of topological sort of a graph that is customized
 * to handle cycles
 */
export function optimizeDepsHierarchy(graph: AdjacencyMap): {
  /**
   * Topologically sorted graph vertices
   */
  sorted: List<string>
  /**
   * Map with data where cycles are detected
   */
  circuitsResolutions: AdjacencyMap
} {
  /**
   * 1. Find components
   * 2. Construct Map<string, Set<string>> with components (key - each vert, value - its component)
   * 3. Sort each component topologically
   * 4. Find circuit resolutions in each component (using sorted components from step 3)
   * 5. Sort initial graph alphabetically -> topologically
   * 6. Return result from 4 and 5
   */

  // 1.
  const components = findStronglyConnectedComponents(graph)

  // 2.
  const componentsMap = buildComponentsMap(components)

  // 3.
  const componentsSorted: Map<Set<string>, List<string>> = components.reduce(
    (map, x) => map.set(x, topologicalSortWithIgnoranceOfCircuits(x, graph)),
    Map(),
  )

  // 4.
  const circuitsResolutions: Map<string, Set<string>> = componentsSorted
    .valueSeq()
    .reduce((map, sortedComponent) => map.merge(findCircuitsResolutions(sortedComponent, graph)), Map())

  // 5.
  const verticesSorted = sortCyclicGraphTopologicallyUsingComponents(graph, componentsMap, componentsSorted)

  return {
    sorted: verticesSorted,
    circuitsResolutions,
  }
}

export function sortCyclicGraphTopologicallyUsingComponents(
  graph: AdjacencyMap,
  componentsMap: AdjacencyMap,
  componentsSorted: Map<Set<String>, List<string>>,
): List<string> {
  const used = Set<string>().asMutable()
  const stack = List<string>().asMutable()

  function dfs(v: string): void {
    const component = componentsMap.get(v)!
    used.merge(component) // itself with all siblings

    for (const u of graph.get(v)!) {
      if (!used.has(u)) dfs(u)
    }

    const sortedComponent = componentsSorted.get(component)!
    stack.concat(sortedComponent)
  }

  for (const v of graph.keySeq().sort()) {
    if (!used.has(v)) dfs(v)
  }

  return stack.asImmutable()
}

export function findCircuitsResolutions(sortedComponent: List<string>, graph: AdjacencyMap): Map<string, Set<string>> {
  const indices = sortedComponent.reduce<Map<string, number>>((map, v, i) => map.set(v, i), Map())

  return sortedComponent.reduce((map, v, i) => {
    return graph.get(v)!.reduce((map, u) => {
      const uIdx = indices.get(u)
      if (typeof uIdx !== 'number' || uIdx < i) return map
      return map.update(v, (set) => (set ?? Set()).add(u))
    }, map)
  }, Map())
}

export function graphToAdjacencyList(graph: AdjacencyMap): {
  adjacencyList: number[][]
  adjacencyListKeys: string[]
} {
  const keys: string[] = Array.from(graph.keySeq().sort().reverse())
  const keysIdx: Map<string, number> = keys.reduce((map, key, idx) => map.set(key, idx), Map<string, number>())
  const list: number[][] = new Array(keys.length)

  for (let i = 0; i < keys.length; i++) {
    list[i] = graph
      .get(keys[i])!
      .map((u) => keysIdx.get(u)!)
      .toArray()
  }

  console.log({ list, keys })

  return { adjacencyList: list, adjacencyListKeys: keys }
}

/**
 * Uses Tarjan's algorithm
 */
export function findStronglyConnectedComponents(graph: AdjacencyMap): Set<Set<string>> {
  const graphTarjan = new Graph()

  for (const [v, deps] of graph.entries()) {
    graphTarjan.add(v, [...deps])
  }

  const components = graphTarjan.getStronglyConnectedComponents()

  return components.reduce((set, arr) => set.add(Set(arr.map((x) => x.name))), Set<Set<string>>())
}

export function topologicalSortWithIgnoranceOfCircuits(component: Set<string>, graph: AdjacencyMap): List<string> {
  const used = Set<string>().asMutable()
  const stack = List<string>().asMutable()

  function dfs(v: string): void {
    used.add(v)
    for (const u of graph.get(v)!) {
      if (component.has(u) && !used.has(u)) dfs(u)
    }
    stack.push(v)
  }

  for (const v of component) {
    if (!used.has(v)) dfs(v)
  }

  return stack.asImmutable()
}

export function buildComponentsMap(components: Set<Set<string>>): Map<string, Set<string>> {
  return components.reduce((map, component) => component.reduce((map, v) => map.set(v, component), map), Map())
}
