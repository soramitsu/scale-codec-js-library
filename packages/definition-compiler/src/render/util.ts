import { Seq } from 'immutable'
import { assert } from '@scale-codec/util'

export const INDENTATION = ' '.repeat(4)

export type ImportOptions = string | { source: string; as: string }

export function renderImports(
  imports: Iterable<ImportOptions>,
  moduleName: string,
  options?: {
    type?: boolean
  },
): string {
  const joined = Seq(imports)
    .map((x) => (typeof x === 'string' ? x : `${x.source} as ${x.as}`))
    .sort()
    .join(',\n' + INDENTATION)

  assert(joined, 'Empty imports are not allowed')

  return `import${options?.type ? ' type' : ''} {\n${INDENTATION}${joined}\n} from '${moduleName}'`
}

export interface DIScope<T> {
  provide: <R>(state: T, fn: () => R) => R
  inject: () => T
}

export function createDIScope<T>(): DIScope<T> {
  const stack: T[] = []

  return {
    provide: (value, fn) => {
      stack.push(value)
      const result = fn()
      stack.pop()
      return result
    },
    inject: () => {
      if (!stack.length) throw new Error('Call `use()` only inside of the `within()` callback')
      return stack[stack.length - 1]
    },
  }
}
