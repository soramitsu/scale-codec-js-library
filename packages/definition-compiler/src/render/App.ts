import { defineComponent, PropType, computed, compile } from 'vue';
import { NamespaceDefinition, TypeDef } from '../definitions';
import { byValue, byString } from 'sort-es';
import { provideConfig } from './config';

import Collector from './components/Collector';
import Preamble from './components/Preamble';
import DefRenderer from './components/DefRenderer';

export default defineComponent({
    components: {
        Collector,
        Preamble,
        DefRenderer,
    },
    props: {
        defmap: {
            type: Object as PropType<NamespaceDefinition>,
            required: true,
        },
        importLib: {
            type: String,
            default: 'sample',
        },
    },
    setup(props) {
        const defsList = computed<{ tyName: String; def: TypeDef }[]>(() => {
            const items = Object.entries(props.defmap);
            items.sort(byValue((x) => x[0], byString()));
            return items.map(([tyName, def]) => ({ tyName, def }));
        });

        provideConfig({
            importLib: props.importLib,
        });

        return {
            defsList,
        };
    },
    render: compile(`
        <collector>
            <preamble />


            <template v-for="x in defsList" :key="x.tyName">
                <w t="\n\n" />
                <def-renderer v-bind="x" />
            </template>
        </collector>
    `),
});
