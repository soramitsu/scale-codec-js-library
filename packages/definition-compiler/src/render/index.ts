import {
    NamespaceDefinition,
    RenderNamespaceDefinitionParams,
    TypeDef,
    DefEnumVariant,
    DefStructField,
} from '../types';
import { Set as SetImmutable } from 'immutable';
import { renderImports, createStateScope } from './util';
import { byValue, byString } from 'sort-es';
import { DefaultAvailableBuilders } from '../const';

function namespaceDefinitionToList(val: NamespaceDefinition): { tyName: string; def: TypeDef }[] {
    const items = Object.entries(val);
    items.sort(byValue((x) => x[0], byString()));
    return items.map(([tyName, def]) => ({ tyName, def }));
}

enum BaseType {
    Instance = 'ScaleInstance',
    Builder = 'ScaleBuilder',
    InstanceViaBuilder = 'InstanceViaBuilder',
    InnerValue = 'InnerValue',
    UnwrappedValue = 'UnwrappedValue',
    Enum = 'Enum',
    Valuable = 'Valuable',
    Option = 'Option',
    Result = 'Result',
}

// =========

interface RenderParams {
    runtimeLib: string;
    runtimeTypes: Set<String>;
    rollupSingleTuples: boolean;
}

const { within: withinRenderParams, use: useRenderParams } = createStateScope<RenderParams>();

// =========

interface ImportsCollector {
    collectRef: (ref: string) => void;
    collectImport: (name: string) => void;
    getRuntimeLibImports: () => Set<string>;
}

function instanceViaBuilder(ref: string): string {
    return `${touchBase(BaseType.InstanceViaBuilder)}<typeof ${touchRef(ref)}>`;
}

function createImportsCollector(): ImportsCollector {
    const { runtimeTypes } = useRenderParams();

    let imports = SetImmutable<string>();

    return {
        getRuntimeLibImports: () => new Set(imports),
        collectRef: (ref) => {
            if (runtimeTypes.has(ref)) {
                imports = imports.add(ref);
            }
        },
        collectImport: (name) => {
            imports = imports.add(name);
        },
    };
}

const { within: withinCollector, use: useCollector } = createStateScope<ImportsCollector>();

// =========

const { within: withinCurrentTyName, use: useCurrentTyName } = createStateScope<string>();

// =========

function renderBuilder(props: { valueTy: string | null; createHelper: string; createHelperArgs: string }): string {
    const ty = useCurrentTyName();
    const helperGeneric = props.valueTy ? `<${props.valueTy}>` : '';

    return `export var ${ty} = ${touchImport(props.createHelper)}${helperGeneric}('${ty}', ${props.createHelperArgs})`;
}

function touchRef(ref: string): string {
    useCollector().collectRef(ref);
    return ref;
}

function touchImport(name: string): string {
    useCollector().collectImport(name);
    return name;
}

function touchBase(ty: BaseType): string {
    useCollector().collectImport(ty);
    return ty;
}

/**
 * ref -> `() => ${ref}`
 */
function refFn(ref: string): string {
    return `() => ${touchRef(ref)}`;
}

function linesJoin(lines: string[], joiner = '\n\n'): string {
    return lines.join(joiner);
}

// =========

function renderAlias(to: string): string {
    return renderBuilder({
        valueTy: [BaseType.InnerValue, BaseType.UnwrappedValue]
            .map((x) => `${touchBase(x)}<typeof ${touchRef(to)}>`)
            .join(', '),
        createHelper: 'createAliasBuilder',
        createHelperArgs: refFn(to),
    });
}

function renderVoidAlias(): string {
    const { runtimeLib } = useRenderParams();

    return renderImport({ nameInModule: 'Void', module: runtimeLib });
}

function renderVec(item: string): string {
    return renderBuilder({
        valueTy: `${instanceViaBuilder(item)}[]`,
        createHelper: 'createVecBuilder',
        createHelperArgs: refFn(item),
    });
}

function renderStruct(fields: DefStructField[]): string {
    if (!fields.length) {
        return renderVoidAlias();
    }

    const valueTypeFields: string[] = fields.map((x) => `${x.name}: ${instanceViaBuilder(x.ref)}`);

    const schemaItems = fields.map((x) => `['${x.name}', ${refFn(x.ref)}]`);
    const schema = `[${schemaItems.join(', ')}]`;

    return renderBuilder({
        valueTy: `{\n    ${valueTypeFields.join(',\n    ')}\n}`,
        createHelper: 'createStructBuilder',
        createHelperArgs: `${schema}`,
    });
}

function renderTuple(refs: string[]): string {
    if (!refs.length) {
        return renderVoidAlias();
    }

    const { rollupSingleTuples } = useRenderParams();
    if (rollupSingleTuples && refs.length === 1) return renderAlias(refs[0]);

    const valueEntries: string[] = refs.map(instanceViaBuilder);
    const codecs: string[] = refs.map(refFn);

    return renderBuilder({
        valueTy: `[\n    ${valueEntries.join(',\n    ')}\n]`,
        createHelper: 'createTupleBuilder',
        createHelperArgs: `[${codecs.join(', ')}]`,
    });
}

function renderEnum(variants: DefEnumVariant[]): string {
    const definitionTyLines: string[] = variants.map((x) => {
        const right = x.ref ? `${touchBase(BaseType.Valuable)}<${instanceViaBuilder(x.ref)}>` : 'null';
        return `${x.name}: ${right}`;
    });

    const schemaLines: string[] = variants.map((x) => {
        const items = [x.discriminant, `'${x.name}'`];
        x.ref && items.push(refFn(x.ref));
        return `[${items.join(', ')}]`;
    });

    return renderBuilder({
        valueTy: `${touchBase(BaseType.Enum)}<{\n    ${definitionTyLines.join(',\n    ')}\n}>`,
        createHelper: 'createEnumBuilder',
        createHelperArgs: `[${schemaLines.join(', ')}]`,
    });
}

function renderSet(item: string): string {
    return renderBuilder({
        valueTy: `Set<${instanceViaBuilder(item)}>`,
        createHelper: 'createSetBuilder',
        createHelperArgs: refFn(item),
    });
}

function renderMap(key: string, value: string): string {
    return renderBuilder({
        valueTy: `Map<${instanceViaBuilder(key)}, ${instanceViaBuilder(value)}>`,
        createHelper: 'createMapBuilder',
        createHelperArgs: [key, value].map(refFn).join(', '),
    });
}

function renderArray(item: string, len: number): string {
    return renderBuilder({
        valueTy: `${instanceViaBuilder(item)}[]`,
        createHelper: `createArrayBuilder`,
        createHelperArgs: `${refFn(item)}, ${len}`,
    });
}

function renderBytesArray(len: number): string {
    return renderBuilder({
        valueTy: null,
        createHelper: 'createBytesArrayBuilder',
        createHelperArgs: `${len}`,
    });
}

function renderOption(some: string): string {
    return renderBuilder({
        valueTy: `${touchBase(BaseType.Option)}<${instanceViaBuilder(some)}>`,
        createHelper: 'createOptionBuilder',
        createHelperArgs: refFn(some),
    });
}

function renderResult(ok: string, err: string): string {
    return renderBuilder({
        valueTy: `${touchBase(BaseType.Result)}<${instanceViaBuilder(ok)}, ${instanceViaBuilder(err)}>`,
        createHelper: 'createResultBuilder',
        createHelperArgs: [ok, err].map(refFn).join(', '),
    });
}

function renderImport({ nameInModule, module: moduleName }: { nameInModule?: string | null; module: string }): string {
    const ty = useCurrentTyName();

    return linesJoin(
        [renderImports([nameInModule ? { source: nameInModule, as: ty } : ty], moduleName), `export { ${ty} }`],
        '\n',
    );
}

function renderParticularDef(tyName: string, def: TypeDef): string {
    const particularRendered: string = withinCurrentTyName(tyName, () => {
        switch (def.t) {
            case 'alias':
                return renderAlias(def.ref);
            case 'vec':
                return renderVec(def.item);
            case 'struct':
                return renderStruct(def.fields);
            case 'tuple':
                return renderTuple(def.items);
            case 'enum':
                return renderEnum(def.variants);
            case 'set':
                return renderSet(def.entry);
            case 'map':
                return renderMap(def.key, def.value);
            case 'array':
                return renderArray(def.item, def.len);
            case 'bytes-array':
                return renderBytesArray(def.len);
            case 'option':
                return renderOption(def.some);
            case 'result':
                return renderResult(def.ok, def.err);
            case 'import':
                return renderImport(def);
            default: {
                const uncovered: never = def;
                throw new Error(`Undefined type definition: ${uncovered}`);
            }
        }
    });

    return particularRendered;
}

function renderPreamble(): string {
    const { runtimeLib } = useRenderParams();
    const { getRuntimeLibImports: getCoreImports } = useCollector();

    const lines = [];

    const imports = getCoreImports();
    if (imports.size) {
        lines.push(renderImports(imports, runtimeLib));
    }

    return linesJoin(lines);
}

/**
 * Renders provided definition into a valid TypeScript code.
 */
export function renderNamespaceDefinition(
    definition: NamespaceDefinition,
    params?: RenderNamespaceDefinitionParams,
): string {
    return withinRenderParams(
        {
            runtimeLib: params?.runtimeLib ?? '@scale-codec/definition-runtime',
            runtimeTypes: params?.runtimeTypes ?? DefaultAvailableBuilders,
            rollupSingleTuples: params?.rollupSingleTuplesIntoAliases ?? false,
        },
        () =>
            withinCollector(createImportsCollector(), () => {
                const renderedTypes = namespaceDefinitionToList(definition)
                    .map(({ tyName, def }) => renderParticularDef(tyName, def))
                    .join('\n\n');

                const preamble = renderPreamble();

                return [preamble, renderedTypes].join('\n\n').trim() + '\n';
            }),
    );
}
