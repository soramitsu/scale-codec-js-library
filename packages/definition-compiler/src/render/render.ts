import { createApp, renderApp } from './vue-code';
import { NamespaceDefinition } from '../definitions';
import App from './App';
import { WithDefPart, AddPartSuffix } from './components/current-def-part';
import { WithCurrentTypeName, TyName } from './components/current-type-name';
import Whitespace from './components/Whitespace';
import { TypeReference, Core } from './components/links';
import TypeExport from './components/TypeExport';
import Fields from './components/Fields';
import UseConfig from './components/UseConfig';
import VoidAlias from './components/VoidAlias';

export interface RenderNamespaceDefinitionParams {
    /**
     * Runtime library with STD codecs + reexports from the core library.
     *
     * @remarks
     * See `@scale-codec/definition-runtime` library - it exists specially for it.
     */
    importLib: string;
    /**
     * Single tuples are always an arrays with a single element. It is possible to make final code cleaner
     * (and a bit performant) if render such tuples just as aliases for the inner element. It is optional feature.
     */
    rollupSingleTuplesIntoAliases?: boolean;
}

/**
 * Renders provided definition into a valid TypeScript code.
 */
export async function renderNamespaceDefinition(
    def: NamespaceDefinition,
    params: RenderNamespaceDefinitionParams,
): Promise<string> {
    const app = createApp(App, {
        defmap: def,
        importLib: params.importLib,
        rollupSingleTuples: params.rollupSingleTuplesIntoAliases ?? false,
    })
        .component('WithDefPart', WithDefPart)
        .component('WithCurrentTypeName', WithCurrentTypeName)
        .component('TyName', TyName)
        .component('Ref', TypeReference)
        .component('Core', Core)
        .component('W', Whitespace)
        .component('Fields', Fields)
        .component('AddPartSuffix', AddPartSuffix)
        .component('UseConfig', UseConfig)
        .component('VoidAlias', VoidAlias)
        .component('Export', TypeExport);

    return renderApp(app);
}
