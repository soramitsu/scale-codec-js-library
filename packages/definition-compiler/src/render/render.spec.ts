import { NamespaceDefinition } from './definitions';
import { renderNamespaceDefinition } from './render';

async function expectRenderToMatchSnapshot(
    def: NamespaceDefinition,
    importLib = '@scale-codec/definition-compiler-runtime',
) {
    expect(await renderNamespaceDefinition(def, { importLib })).toMatchSnapshot();
}

it('Renders vec', () => {
    return expectRenderToMatchSnapshot({
        Vec_str: {
            t: 'vec',
            item: 'str',
        },
    });
});

it('Renders struct + tuple', () => {
    return expectRenderToMatchSnapshot({
        Structural: {
            t: 'struct',
            fields: [
                {
                    name: 'nums',
                    ref: 'Tuple_u8_i128',
                },
                {
                    name: 'mur',
                    ref: 'bool',
                },
            ],
        },
        Tuple_u8_i128: {
            t: 'tuple',
            items: ['u8', 'i128'],
        },
    });
});

it('Renders enum', () => {
    return expectRenderToMatchSnapshot({
        Message: {
            t: 'enum',
            variants: [
                {
                    name: 'Quit',
                    discriminant: 0,
                },
                {
                    name: 'Greeting',
                    discriminant: 1,
                    ref: 'str',
                },
            ],
        },
    });
});

it('Renders set', () => {
    return expectRenderToMatchSnapshot({
        Set_Message: {
            t: 'set',
            entry: 'Message',
        },
    });
});

it('Renders map', () => {
    return expectRenderToMatchSnapshot({
        Map_str_Message: {
            t: 'map',
            key: 'str',
            value: 'Message',
        },
    });
});

it('Renders array', () => {
    return expectRenderToMatchSnapshot({
        Array_Item_15: {
            t: 'array',
            item: 'Item',
            len: 15,
        },
    });
});

it('Renders bytes array', () => {
    return expectRenderToMatchSnapshot({
        Hash: {
            t: 'bytes-array',
            len: 64,
        },
    });
});

it('Renders option', () => {
    return expectRenderToMatchSnapshot({
        OptionHash: {
            t: 'option',
            some: 'Hash',
        },
    });
});
