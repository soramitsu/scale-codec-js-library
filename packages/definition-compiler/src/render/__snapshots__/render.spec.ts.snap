// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders array 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  decodeArray,
  encodeArray,
  makeEncoderAsIsRespectable,
} from '@scale-codec/definition-compiler-runtime'

// Array_Item_15

export type Array_Item_15_Decoded = Item_Decoded[]

export type Array_Item_15_Encodable = (Item_Encodable | EncodeAsIs)[]

const Array_Item_15_item_encode = makeEncoderAsIsRespectable(Item_encode)
const Array_Item_15_len = 15

export function Array_Item_15_decode(bytes: Uint8Array): DecodeResult<Array_Item_15_Decoded> {
  return decodeArray(bytes, Item_decode, Array_Item_15_len)
}

export function Array_Item_15_encode(encodable: Array_Item_15_Encodable): Uint8Array {
  return encodeArray(encodable, Array_Item_15_item_encode, Array_Item_15_len)
}
"
`;

exports[`Renders bytes array 1`] = `
"/* eslint-disable */
import { DecodeResult, decodeUint8Array, encodeUint8Array } from '@scale-codec/definition-compiler-runtime'

// Hash

export type Hash_Decoded = Uint8Array

export type Hash_Encodable = Uint8Array

const Hash_len = 64

export function Hash_decode(bytes: Uint8Array): DecodeResult<Hash_Decoded> {
  return decodeUint8Array(bytes, Hash_len)
}

export function Hash_encode(encodable: Hash_Encodable): Uint8Array {
  return encodeUint8Array(encodable, Hash_len)
}
"
`;

exports[`Renders empty struct as void alias 1`] = `
"/* eslint-disable */

// EmptyStruct

// EmptyStruct is just a void alias

import {
  Void_Decoded as EmptyStruct_Decoded,
  Void_Encodable as EmptyStruct_Encodable,
  Void_decode as EmptyStruct_decode,
  Void_encode as EmptyStruct_encode,
} from '@scale-codec/definition-compiler-runtime'

export { EmptyStruct_Decoded, EmptyStruct_Encodable, EmptyStruct_decode, EmptyStruct_encode }
"
`;

exports[`Renders empty tuple as void alias 1`] = `
"/* eslint-disable */

// EmptyTuple

// EmptyTuple is just a void alias

import {
  Void_Decoded as EmptyTuple_Decoded,
  Void_Encodable as EmptyTuple_Encodable,
  Void_decode as EmptyTuple_decode,
  Void_encode as EmptyTuple_encode,
} from '@scale-codec/definition-compiler-runtime'

export { EmptyTuple_Decoded, EmptyTuple_Encodable, EmptyTuple_decode, EmptyTuple_encode }
"
`;

exports[`Renders enum 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  Enum,
  EnumDecoders,
  EnumEncoders,
  Valuable,
  decodeEnum,
  encodeEnum,
  makeEncoderAsIsRespectable,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
} from '@scale-codec/definition-compiler-runtime'

// Message

export type Message_Decoded = Enum<{
  Quit: null
  Greeting: Valuable<str_Decoded>
}>

export type Message_Encodable = Enum<{
  Quit: null
  Greeting: Valuable<str_Encodable | EncodeAsIs>
}>

// Message enum tools

const Message_decoders: EnumDecoders = {
  0: { v: 'Quit' },
  1: { v: 'Greeting', decode: str_decode },
}
const Message_encoders: EnumEncoders = {
  Quit: { d: 0 },
  Greeting: { d: 1, encode: makeEncoderAsIsRespectable(str_encode) },
}

// Message tools end

export function Message_decode(bytes: Uint8Array): DecodeResult<Message_Decoded> {
  return decodeEnum(bytes, Message_decoders)
}

export function Message_encode(encodable: Message_Encodable): Uint8Array {
  return encodeEnum(encodable, Message_encoders)
}
"
`;

exports[`Renders map 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  Encode,
  EncodeAsIs,
  decodeMap,
  encodeMap,
  makeEncoderAsIsRespectable,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
} from '@scale-codec/definition-compiler-runtime'

// Map_str_Message

export type Map_str_Message_Decoded = Map<str_Decoded, Message_Decoded>

export type Map_str_Message_Encodable = Map<str_Encodable | EncodeAsIs, Message_Encodable | EncodeAsIs>

const [Map_str_Message_encode_key, Map_str_Message_encode_value] = ([str_encode, Message_encode] as any).map(
  makeEncoderAsIsRespectable
) as [Encode<str_Encodable | EncodeAsIs>, Encode<Message_Encodable | EncodeAsIs>]

export function Map_str_Message_decode(bytes: Uint8Array): DecodeResult<Map_str_Message_Decoded> {
  return decodeMap(bytes, str_decode, Message_decode)
}

export function Map_str_Message_encode(encodable: Map_str_Message_Encodable): Uint8Array {
  return encodeMap(encodable, Map_str_Message_encode_key, Map_str_Message_encode_value)
}
"
`;

exports[`Renders option 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  EnumDecoders,
  EnumEncoders,
  Option,
  decodeEnum,
  encodeEnum,
  makeEncoderAsIsRespectable,
} from '@scale-codec/definition-compiler-runtime'

// OptionHash

export type OptionHash_Decoded = Option<Hash_Decoded>

export type OptionHash_Encodable = Option<Hash_Encodable | EncodeAsIs>

const OptionHash_decoders: EnumDecoders = {
  0: { v: 'None' },
  1: { v: 'Some', decode: Hash_decode },
}
const OptionHash_encoders: EnumEncoders = {
  None: { d: 0 },
  Some: { d: 1, encode: makeEncoderAsIsRespectable(Hash_encode) },
}

export function OptionHash_decode(bytes: Uint8Array): DecodeResult<OptionHash_Decoded> {
  return decodeEnum(bytes, OptionHash_decoders)
}

export function OptionHash_encode(encodable: OptionHash_Encodable): Uint8Array {
  return encodeEnum(encodable, OptionHash_encoders)
}
"
`;

exports[`Renders set 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  decodeSet,
  encodeSet,
  makeEncoderAsIsRespectable,
} from '@scale-codec/definition-compiler-runtime'

// Set_Message

export type Set_Message_Decoded = Set<Message_Decoded>

export type Set_Message_Encodable = Set<Message_Encodable | EncodeAsIs>

// Set_Message set tools

const Set_Message_entry_encode = makeEncoderAsIsRespectable(Message_encode)

// Set_Message tools end

export function Set_Message_decode(bytes: Uint8Array): DecodeResult<Set_Message_Decoded> {
  return decodeSet(bytes, Message_decode)
}

export function Set_Message_encode(encodable: Set_Message_Encodable): Uint8Array {
  return encodeSet(encodable, Set_Message_entry_encode)
}
"
`;

exports[`Renders struct + tuple 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  bool_Decoded,
  bool_Encodable,
  bool_decode,
  bool_encode,
  decodeStruct,
  decodeTuple,
  encodeStruct,
  encodeTuple,
  i128_Decoded,
  i128_Encodable,
  i128_decode,
  i128_encode,
  makeEncoderAsIsRespectable,
  u8_Decoded,
  u8_Encodable,
  u8_decode,
  u8_encode,
} from '@scale-codec/definition-compiler-runtime'

// Structural

export type Structural_Decoded = {
  nums: Tuple_u8_i128_Decoded
  mur: bool_Decoded
}

export type Structural_Encodable = {
  nums: Tuple_u8_i128_Encodable | EncodeAsIs
  mur: bool_Encodable | EncodeAsIs
}

const Structural_order: (keyof Structural_Decoded)[] = ['nums', 'mur']
const Structural_decoders = {
  nums: Tuple_u8_i128_decode,
  mur: bool_decode,
}
const Structural_encoders = {
  nums: makeEncoderAsIsRespectable(Tuple_u8_i128_encode),
  mur: makeEncoderAsIsRespectable(bool_encode),
}

export function Structural_decode(bytes: Uint8Array): DecodeResult<Structural_Decoded> {
  return decodeStruct(bytes, Structural_decoders, Structural_order)
}

export function Structural_encode(encodable: Structural_Encodable): Uint8Array {
  return encodeStruct(encodable, Structural_encoders, Structural_order)
}

// Tuple_u8_i128

export type Tuple_u8_i128_Decoded = [u8_Decoded, i128_Decoded]

export type Tuple_u8_i128_Encodable = [u8_Encodable | EncodeAsIs, i128_Encodable | EncodeAsIs]

// Tuple_u8_i128 tuple-related tools

const Tuple_u8_i128_decoders = [u8_decode, i128_decode]
const Tuple_u8_i128_encoders = ([u8_encode, i128_encode] as any).map(makeEncoderAsIsRespectable)

// Tuple_u8_i128 tools end

export function Tuple_u8_i128_decode(bytes: Uint8Array): DecodeResult<Tuple_u8_i128_Decoded> {
  return decodeTuple(bytes, Tuple_u8_i128_decoders as any)
}

export function Tuple_u8_i128_encode(encodable: Tuple_u8_i128_Encodable): Uint8Array {
  return encodeTuple(encodable, Tuple_u8_i128_encoders as any)
}
"
`;

exports[`Renders vec 1`] = `
"/* eslint-disable */
import {
  DecodeResult,
  EncodeAsIs,
  decodeVec,
  encodeVec,
  makeEncoderAsIsRespectable,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
} from '@scale-codec/definition-compiler-runtime'

// Vec_str

export type Vec_str_Decoded = str_Decoded[]

export type Vec_str_Encodable = (str_Encodable | EncodeAsIs)[]

const Vec_str_item_encode = makeEncoderAsIsRespectable(str_encode)

export function Vec_str_decode(bytes: Uint8Array): DecodeResult<Vec_str_Decoded> {
  return decodeVec(bytes, str_decode)
}

export function Vec_str_encode(encodable: Vec_str_Encodable): Uint8Array {
  return encodeVec(encodable, Vec_str_item_encode)
}
"
`;
